lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
-            auto iter = std::find(dots.begin(), dots.end(), tempInAll);
-            if (iter != dots.end())
-              continue;
-            auto db = getBlockNumInFor(tempInAll, forOp);
-            auto fb = getBlockNumInFor(firstUse, forOp);
-            if (db < fb ||
-                (db == fb && db >= 0 && tempInAll->isBeforeInBlock(firstUse)))
-              hasSyncDot = true;
-          }
-          auto CArg = dotOp.getOperand(2);
-          if (!(selfDirectDepend ||
-                (depend && !selfDirectDepend && hasSyncDot)) ||
-              !CArg.hasOneUse())
-            valid = false;
-
-          // Both operands needs to have been multi-buffered if they come from
-          // shared memory otherwise we will have a race condition as the shared
-          // memory will be overriden before the wgmma wait.
-          // Note that we relay on the assumption that `createAsynOps` will have
-          // allocated enough buffers to pipeline dot operation with 2 stages.
-          for (Value operand : {dotOp.getOperand(0), dotOp.getOperand(1)}) {
-            auto operandEncoding =
-                operand.getType().cast<TensorOrMemDesc>().getEncoding();
-            if (!operandEncoding.isa<ttg::SharedEncodingAttr>())
-              continue;
-            Value transitiveOperand = operand;
-            while (isa_and_nonnull<ttg::ConvertLayoutOp, tt::TransOp>(
-                transitiveOperand.getDefiningOp()))
-              transitiveOperand =
-                  transitiveOperand.getDefiningOp()->getOperand(0);
-            if (forOp.isDefinedOutsideOfLoop(transitiveOperand))
-              continue;
-            if (transitiveOperand.getDefiningOp<ttg::MemDescSubviewOp>() ==
-                nullptr)
-              valid = false;
-          }
 
-          if (valid) {
-            dots.push_back(dotOp);
-            resultNeedSync.push_back(
-                dotOp->getUses().begin()->getOperandNumber());
-          }
-        }
-      }
+  // If we have a synchyronizing dot, convert it to async and insert a wait
+  // right after it.  Otherwise, we'll insert the wait at the end of the loop.
+  IRRewriter builder(forOp.getContext());
+  if (synchronizingDot) {
+    builder.setInsertionPointAfter(synchronizingDot);
+    builder.replaceOpWithNewOp<ttng::DotAsyncOp>(
+        synchronizingDot.getOperation(), synchronizingDot.getA(),
+        synchronizingDot.getB(), synchronizingDot.getC(),
+        synchronizingDot.getAllowTF32(),
+        synchronizingDot.getMaxNumImpreciseAcc());
+  } else {
+    builder.setInsertionPoint(forOp.getBody()->getTerminator());
+  }
+
+  SmallVector<Value> waitOperands;
+  for (auto [asyncDot, iterArgIdx] : asyncDots) {
+    Value dotResult = syncAtBeginning
+                          ? cast<Value>(forOp.getRegionIterArg(iterArgIdx))
+                          : cast<Value>(asyncDot->getResult(0));
+    if (!dotResult.use_empty())
+      waitOperands.push_back(dotResult);
+  }
```

Comment:
Is inserting the wait at the end of the loop correct? That wasn't the behavior of previous code right?

Reply:
Great point, fixed.